### PR TITLE
Degrade Calico status on typha autoscale issues

### PR DESCRIPTION
**Pull request by Brian-McM**
This commit ties the Calico tigera status to the typha autoscalar, degrading the status if there's an issue.

An error check that has been added is to make sure that there are enough linux nodes available to schedule the required typha pods on, and now if there's not, an error is logged and the Calico tigera status is set to degraded

If the auto scalar degraded the status, the core controller will see it if it reconciles and trigger the autoscalar run immediately. If this run fails the controller will set the degraded status again and try to re reconcile later. Note the the ticker in typha is reset if the run is triggered by the controller.

## For PR author

- [x] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [x] Milestone set according to targeted release.
- [x] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/controller/installation/typha_autoscaler_test.go
Comment (tmjd) [inline pkg/controller/installation/typha_autoscaler_test.go:131]: Instead of using a sleep here it should work just to do `ta.isDegraded()`, that would ensure that autoscale was ran and SetDegraded would be set if degraded.

File: pkg/controller/installation/typha_autoscaler.go
Comment (tmjd) [inline pkg/controller/installation/typha_autoscaler.go:78]: Should we really be making these channels with 100 buffer? It should be fine to be 1 for both.

**Review by tmjd — COMMENTED** [review]
My thoughts

File: pkg/controller/installation/core_controller.go
Comment (Brian-McM) [reply pkg/controller/installation/core_controller.go:563]: Ah right! You would think I'd remember this as I've been burned by it so many times lol.

File: pkg/controller/installation/typha_autoscaler.go
Comment (Brian-McM) [reply pkg/controller/installation/typha_autoscaler.go:151]: My reasoning was to make sure that it doesn't try to immediately autoscale again if it was triggered manually. If there was 5 seconds left on the ticker and this was triggered manually the autoscale would be kicked off again almost immediately.

It's not _that_ big a deal if we don't reset the ticker, but on the other hand it's really easy to do so I figured why not.

File: pkg/controller/installation/typha_autoscaler_test.go
Comment (Brian-McM) [reply pkg/controller/installation/typha_autoscaler_test.go:131]: Technically there's still a race there, although it's a bit smaller. The initial autoscale run is done in the go routine, so the degraded check could theoretically beat the run.

I think putting this to a milli second then doing the degraded check would be good, just to ensure this. WDYT?

File: pkg/controller/installation/typha_autoscaler.go
Comment (Brian-McM) [reply pkg/controller/installation/typha_autoscaler.go:78]: Meh, probably not. I guess we don't need to, the go routine is essentially locked while triggerRun and isDegraded are being called anyway.

Comment (Brian-McM) [reply pkg/controller/installation/typha_autoscaler.go]: I set it to 0, which means it blocks until the object is read off of the channel. I think that's probably what we want.

File: pkg/controller/installation/typha_autoscaler.go
Comment (tmjd) [reply pkg/controller/installation/typha_autoscaler.go:151]: Yeah I guess getting the list of nodes might be significant in a large cluster. So I'm good with keeping it.

File: pkg/controller/installation/typha_autoscaler_test.go
Comment (tmjd) [reply pkg/controller/installation/typha_autoscaler_test.go]: The ta.isDegraded() will block until the loop and select happens once and writes the degraded status back. Before the loop even runs we do an autoscale and set degraded so I don't think there is a race here.

**Review by tmjd — APPROVED** [review]
LGTM

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and delete the branch after I've merged it.

**Comment by marvin-tigera** [comment]
I'm sorry but I failed delete the branch after merging the pull request.

**Review by caseydavenport — COMMENTED** [review]
Whoops, never submitted this. It's only one nitty comment though :D

**Pull request merged**
